### PR TITLE
make seeding resources work for scripts in unit groups

### PR DESCRIPTION
**Pull request by davidsbailey**
## Background

We have a bit of of a chicken-and-egg problem for seeding resources. The scenario assumes that `ScriptSeed.seed_from_json` has already been added to `rake seed:scripts`, as described in [PLAT-560].

Scenario:
* resources need a `course_version_id` in order to be seeded
* `rake seed:scripts` runs before `rake seed:courses`: https://github.com/code-dot-org/code-dot-org/blob/b6b6747fbdeab9df23cb0791f62155b737a8392f/dashboard/lib/tasks/seed.rake#L405
* the inclusion of a `Script` in a `UnitGroup` is established during `rake seed:courses`

Therefore, when seeding is run for the first time in an environment, we will not be able to determine the `course_version_id` of a `Script` which is in a `UnitGroup`, and we won't be able to seed resources.

## Terminology

we only got part way through renaming course -> unit group, in particular `rake seed:courses` and `.course` files still use the word "course". Conceptually, "course" and "unit group" are synonyms. I've tried to use the names as they appear in github to minimize confusion.

we want to rename "script" to "unit", but we haven't done that yet, so we're still calling it "script".

## Description

This PR makes seeding resources work for `Script`s in `UnitGroup`s. It deals with the chicken-and-egg problem as follows:
1. the first time we seed a `Script` in a `UnitGroup` on a particular machine, we won't know its course version. `ScriptSeed.seed_from_json` succeeds without importing resources, and prints a warning.
2. `rake seed:courses` then creates `UnitGroup`s and connects them to `Script`s. at this point, the script has a course version, but we have already passed the step for seeding resources.
3. the first deploy to the new machine succeeds, but without populating any resources into the DB. now we are in a bad state:

    * resource links (not yet implemented) won't be renderable on the lesson plan page
    * anything else we can think of to check for?

4. the next time the machine is deployed, the `Script`'s `CourseVersion` is available, and resources are seeded properly. now we are back in a good state.

This is a bit messy, so I am open to other short term options.

one ideas for a long term solution is to split the script seed step into two steps, one which creates an empty script, and one which imports its contents. e.g. `FULL_SEED_TASKS = [..., :create_scripts, :courses, :load_script_contents, ...]`. I don't think we need to decide this now, but it could be good to hear any other options which come to mind while we are thinking about it.

## Testing story

unit tests cover the following scenarios:
1. a `Script` with resources is seeded before its `UnitGroup` is seeded. seeding succeeds without importing resources
2. a `Script` with resources and `UnitGroup` is seeded. seeding succeeds, importing all resources.

## Caveats

this implementation theoretically leaves the following loophole:
1. resources added to script in unit group
2. script and unit group are serialized
3. .course file for that unit group is somehow lost before the unit group is seeded into another environment
4. the script will succeed on every deploy, without loading any resources.

in this case, a warning message will be printed on every deploy, but it may never be visible as long as the seed process keeps succeeding. I'm not sure this is very likely to ever happen, but I wanted to capture it as a downside of this approach anyway.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

[PLAT-560]: https://codedotorg.atlassian.net/browse/PLAT-560

### CODE REVIEW

File: dashboard/lib/services/script_seed.rb [hunk 1]
```diff
@@ -145,8 +145,12 @@ def self.seed_from_json(json_string)
 
         seed_context.script = import_script(script_data)
 
-        # course version must be set before resources are imported.
-        CourseOffering.add_course_offering(seed_context.script)
+        # Course version must be set before resources are imported. If the
+        # script is in a unit group, we must wait and let the next seed step set
+        # the course version on the unit group before resources can be imported.
+        if seed_context.script.is_course
```
Comment (bethanyaconnor) [inline dashboard/lib/services/script_seed.rb:151]: Does `rake seed:courses` add the course offering to a script?

Comment (davidsbailey) [reply dashboard/lib/services/script_seed.rb:151]: kind of. `rake seed:courses` adds the course offering to a unit group. after that, a call to script.get_course_version will navigate through that unit group to find the course version (and, if it wanted, the course offering).
* https://github.com/code-dot-org/code-dot-org/blob/d01e89082009cad3c14055da5c6f9c5ac66038f3/dashboard/lib/tasks/seed.rake#L216-L218
* https://github.com/code-dot-org/code-dot-org/blob/75c067c90a76c5f8bfff6c99329a0f621e5069b8/dashboard/app/models/unit_group.rb#L92

Comment (bethanyaconnor) [reply dashboard/lib/services/script_seed.rb:151]: Ahh makes sense. So this branch was never needed for scripts in a unit group?

Comment (davidsbailey) [reply dashboard/lib/services/script_seed.rb:151]: it's more that this codepath was probably totally broken for scripts in a unit group before this PR. I had only been testing on scripts which were standalone courses i.e. `is_course`.

**Review by bethanyaconnor — APPROVED** [review]
Thanks for thinking through this scenario!

**Pull request merged**